cmake error for v3.6.3 and v3.6.4

Hi!

I am trying to build cuda backend inside a docker container which is build with Dockerfile
```
FROM nvidia/cuda:9.2-cudnn7-devel-ubuntu16.04

RUN APT_INSTALL="apt-get install -y --no-install-recommends" && \
    rm -rf /var/lib/apt/lists/* \
           /etc/apt/sources.list.d/cuda.list \
           /etc/apt/sources.list.d/nvidia-ml.list && \
    apt-get update && \
    DEBIAN_FRONTEND=noninteractive $APT_INSTALL \
        build-essential \
        ca-certificates \
        cmake \
        wget \
        git \
        vim \
        emacs \
        nano \
        htop \
        g++ \
        # ssh for OpenMPI
        openssh-server openssh-client \
        # OpenMPI
        libopenmpi-dev libomp-dev \
        # nccl: for flashlight
        libnccl2 libnccl-dev
``` 

Inside this container I am running
```
cd /tmp && git clone --recursive https://github.com/arrayfire/arrayfire.git && \
cd arrayfire && git checkout v3.6.3 && git submodule update --init --recursive && \
mkdir build && cd build && \
CXXFLAGS=-DOS_LNX cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CPU=OFF -DAF_BUILD_OPENCL=OFF -DAF_BUILD_EXAMPLES=OFF && \
make -j8
```

and getting this error during build

```
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for CL_VERSION_2_0
-- Looking for CL_VERSION_2_0 - not found
-- Looking for CL_VERSION_1_2
-- Looking for CL_VERSION_1_2 - found
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found version "1.2")
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found suitable version "9.2", minimum required is "7.0")
-- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found suitable version "1.2", minimum required is "1.2")
-- Could NOT find OpenGL (missing:  OPENGL_gl_LIBRARY OPENGL_INCLUDE_DIR)
-- Found OpenMP_C: -fopenmp (found version "4.0")
-- Found OpenMP_CXX: -fopenmp (found version "4.0")
-- Could NOT find FreeImage (missing:  FreeImage_INCLUDE_DIR FreeImage_LINK_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find FFTW (missing:  FFTW_INCLUDE_DIR FFTW_LIBRARY FFTWF_LIBRARY)
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Looking for sys/types.h
-- Looking for sys/types.h - found
-- Looking for stdint.h
-- Looking for stdint.h - found
-- Looking for stddef.h
-- Looking for stddef.h - found
-- Check size of void*
-- Check size of void* - done
-- Checking for [Accelerate]
-- Checking for [vecLib]
-- Checking for [cblas - atlas]
-- Checking for [openblas]
-- Checking for [cblas]
-- Checking for [blas]
-- CBLAS library not found.
-- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
-- Could NOT find LAPACK (missing:  LAPACK_INCLUDE_DIR LAPACK_LIBRARIES)
-- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
-- Check size of int
-- Check size of int - done
-- Could NOT find MKL (missing:  MKL_INCLUDE_DIR MKL_Core_LINK_LIBRARY)
-- Could NOT find Boost
-- BOOST_COMPUTE: /tmp/arrayfire/build/boost_compute-prefix/src/boost_compute
-- CUDA_architecture_build_targets: 6.0 6.0
-- Performing Test has_ignored_attributes_flag
-- Performing Test has_ignored_attributes_flag - Success
-- Could NOT find PythonInterp (missing:  PYTHON_EXECUTABLE)
-- Check if compiler accepts -pthread
-- Check if compiler accepts -pthread - yes
-- Configuring done
...
-- Generating done
-- Build files have been written to: /tmp/arrayfire/build
[  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o
[  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o
[  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o
Scanning dependencies of target boost_compute
Scanning dependencies of target gtest
Scanning dependencies of target bin2cpp
[  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o
Scanning dependencies of target af
[  0%] Creating directories for 'boost_compute'
[  0%] Building CXX object CMakeFiles/bin2cpp.dir/CMakeModules/bin2cpp.cpp.o
[  0%] Building CXX object test/gtest/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
[  1%] Performing download step (download, verify and extract) for 'boost_compute'
-- downloading...
     src='https://github.com/boostorg/compute/archive/boost-1.61.0.tar.gz'
     dst='/tmp/arrayfire/build/boost_compute-prefix/src/boost-1.61.0.tar.gz'
     timeout='none'
[  1%] Building CXX object src/api/unified/CMakeFiles/af.dir/algorithm.cpp.o
In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
                 from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
                 from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_float_0.cu:10:
/tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
compilation terminated.
In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
                 from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
                 from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_double_0.cu:10:
/tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
compilation terminated.
In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
                 from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
                 from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_uchar_1.cu:10:
/tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory                                                                                     [287/1873]
compilation terminated.
In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
                 from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
                 from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
                 from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_float_1.cu:10:
/tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
compilation terminated.
CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o.cmake:207 (message):
  Error generating
  /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o


CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o.cmake:207 (message):
  Error generating
  /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o


CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o.cmake:207 (message):
  Error generating
  /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o


CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o.cmake:207 (message):
  Error generating
  /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o


src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:63: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
mpl_float_0.cu.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o] Error 1
make[2]: *** Waiting for unfinished jobs....
src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:77: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
mpl_double_0.cu.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o] Error 1
src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:70: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
mpl_float_1.cu.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o] Error 1
src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:196: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key$
impl_uchar_1.cu.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o] Error 1
CMakeFiles/Makefile2:1331: recipe for target 'src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/all' failed
make[1]: *** [src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
...
[ 12%] Building CXX object src/api/unified/CMakeFiles/af.dir/__/cpp/wrap.cpp.o
[ 12%] Building CXX object src/api/unified/CMakeFiles/af.dir/__/cpp/ycbcr_rgb.cpp.o
[ 12%] Linking CXX shared library libaf.so
[ 12%] Built target af
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

If I am doing the same but for `v3.6.4`:
```
Scanning dependencies of target afcuda
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/kernel/convolve_separable.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/transpose.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/Array.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/blas.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/convolve.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cufft.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cublas.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cusolverDn.cpp.o
[ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cusparse.cpp.o
[ 53%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o
In file included from /tmp/arrayfire/src/backend/cuda/device_manager.cpp:31:0:
/usr/local/cuda/include/cuda_gl_interop.h:67:19: fatal error: GL/gl.h: No such file or directory
compilation terminated.
src/backend/cuda/CMakeFiles/afcuda.dir/build.make:782: recipe for target 'src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o' failed
make[2]: *** [src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:1403: recipe for target 'src/backend/cuda/CMakeFiles/afcuda.dir/all' failed
make[1]: *** [src/backend/cuda/CMakeFiles/afcuda.dir/all] Error 2
Makefile:160: recipe for target 'all' failed
make: *** [all] Error 2
```

Thanks for helping!

Hi Tatiana,

With 3.6.3, I think it is some header inclusion related problem in cmake. I
will check this on a Ubuntu 16.04 system. What is the cmake version you
have on this image?

With 3.6.4 , we reorganized graphics related headers. GL.h not found just
means, you don't have the mesa library installed in your container. Please
note that it is not a requirement to use ArrayFire but to build ArrayFire
you need gl development library installed at the moment.

https://github.com/arrayfire/arrayfire/wiki/Build-Instructions-for-Linux

Regards,
Pradeep.

On Tue 21 May, 2019, 05:48 Tatiana Likhomanenko, <notifications@github.com>
wrote:

> Hi!
>
> I am trying to build cuda backend inside a docker container which is build
> with Dockerfile
>
> FROM nvidia/cuda:9.2-cudnn7-devel-ubuntu16.04
>
> RUN APT_INSTALL="apt-get install -y --no-install-recommends" && \
>     rm -rf /var/lib/apt/lists/* \
>            /etc/apt/sources.list.d/cuda.list \
>            /etc/apt/sources.list.d/nvidia-ml.list && \
>     apt-get update && \
>     DEBIAN_FRONTEND=noninteractive $APT_INSTALL \
>         build-essential \
>         ca-certificates \
>         cmake \
>         wget \
>         git \
>         vim \
>         emacs \
>         nano \
>         htop \
>         g++ \
>         # ssh for OpenMPI
>         openssh-server openssh-client \
>         # OpenMPI
>         libopenmpi-dev libomp-dev \
>         # nccl: for flashlight
>         libnccl2 libnccl-dev
>
> Inside this container I am running
>
> cd /tmp && git clone --recursive https://github.com/arrayfire/arrayfire.git && \
> cd arrayfire && git checkout v3.6.3 && git submodule update --init --recursive && \
> mkdir build && cd build && \
> CXXFLAGS=-DOS_LNX cmake .. -DCMAKE_BUILD_TYPE=Release -DAF_BUILD_CPU=OFF -DAF_BUILD_OPENCL=OFF -DAF_BUILD_EXAMPLES=OFF && \
> make -j8
>
> and getting this error during build
>
> -- The C compiler identification is GNU 5.4.0
> -- The CXX compiler identification is GNU 5.4.0
> -- Check for working C compiler: /usr/bin/cc
> -- Check for working C compiler: /usr/bin/cc -- works
> -- Detecting C compiler ABI info
> -- Detecting C compiler ABI info - done
> -- Detecting C compile features
> -- Detecting C compile features - done
> -- Check for working CXX compiler: /usr/bin/c++
> -- Check for working CXX compiler: /usr/bin/c++ -- works
> -- Detecting CXX compiler ABI info
> -- Detecting CXX compiler ABI info - done
> -- Detecting CXX compile features
> -- Detecting CXX compile features - done
> -- Looking for CL_VERSION_2_0
> -- Looking for CL_VERSION_2_0 - not found
> -- Looking for CL_VERSION_1_2
> -- Looking for CL_VERSION_1_2 - found
> -- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found version "1.2")
> -- Looking for pthread.h
> -- Looking for pthread.h - found
> -- Looking for pthread_create
> -- Looking for pthread_create - not found
> -- Looking for pthread_create in pthreads
> -- Looking for pthread_create in pthreads - not found
> -- Looking for pthread_create in pthread
> -- Looking for pthread_create in pthread - found
> -- Found Threads: TRUE
> -- Found CUDA: /usr/local/cuda (found suitable version "9.2", minimum required is "7.0")
> -- Could NOT find OpenCL (missing:  OpenCL_LIBRARY) (found suitable version "1.2", minimum required is "1.2")
> -- Could NOT find OpenGL (missing:  OPENGL_gl_LIBRARY OPENGL_INCLUDE_DIR)
> -- Found OpenMP_C: -fopenmp (found version "4.0")
> -- Found OpenMP_CXX: -fopenmp (found version "4.0")
> -- Could NOT find FreeImage (missing:  FreeImage_INCLUDE_DIR FreeImage_LINK_LIBRARY)
> -- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
> -- Could NOT find FFTW (missing:  FFTW_INCLUDE_DIR FFTW_LIBRARY FFTWF_LIBRARY)
> -- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
> -- Looking for sys/types.h
> -- Looking for sys/types.h - found
> -- Looking for stdint.h
> -- Looking for stdint.h - found
> -- Looking for stddef.h
> -- Looking for stddef.h - found
> -- Check size of void*
> -- Check size of void* - done
> -- Checking for [Accelerate]
> -- Checking for [vecLib]
> -- Checking for [cblas - atlas]
> -- Checking for [openblas]
> -- Checking for [cblas]
> -- Checking for [blas]
> -- CBLAS library not found.
> -- Could NOT find PkgConfig (missing:  PKG_CONFIG_EXECUTABLE)
> -- Could NOT find LAPACK (missing:  LAPACK_INCLUDE_DIR LAPACK_LIBRARIES)
> -- Could NOT find Doxygen (missing:  DOXYGEN_EXECUTABLE)
> -- Check size of int
> -- Check size of int - done
> -- Could NOT find MKL (missing:  MKL_INCLUDE_DIR MKL_Core_LINK_LIBRARY)
> -- Could NOT find Boost
> -- BOOST_COMPUTE: /tmp/arrayfire/build/boost_compute-prefix/src/boost_compute
> -- CUDA_architecture_build_targets: 6.0 6.0
> -- Performing Test has_ignored_attributes_flag
> -- Performing Test has_ignored_attributes_flag - Success
> -- Could NOT find PythonInterp (missing:  PYTHON_EXECUTABLE)
> -- Check if compiler accepts -pthread
> -- Check if compiler accepts -pthread - yes
> -- Configuring done
> ...
> -- Generating done
> -- Build files have been written to: /tmp/arrayfire/build
> [  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o
> [  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o
> [  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o
> Scanning dependencies of target boost_compute
> Scanning dependencies of target gtest
> Scanning dependencies of target bin2cpp
> [  0%] Building NVCC (Device) object src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o
> Scanning dependencies of target af
> [  0%] Creating directories for 'boost_compute'
> [  0%] Building CXX object CMakeFiles/bin2cpp.dir/CMakeModules/bin2cpp.cpp.o
> [  0%] Building CXX object test/gtest/googletest/CMakeFiles/gtest.dir/src/gtest-all.cc.o
> [  1%] Performing download step (download, verify and extract) for 'boost_compute'
> -- downloading...
>      src='https://github.com/boostorg/compute/archive/boost-1.61.0.tar.gz'
>      dst='/tmp/arrayfire/build/boost_compute-prefix/src/boost-1.61.0.tar.gz'
>      timeout='none'
> [  1%] Building CXX object src/api/unified/CMakeFiles/af.dir/algorithm.cpp.o
> In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
>                  from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
>                  from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_float_0.cu:10:
> /tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
> compilation terminated.
> In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
>                  from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
>                  from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_double_0.cu:10:
> /tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
> compilation terminated.
> In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
>                  from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
>                  from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_uchar_1.cu:10:
> /tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory                                                                                     [287/1873]
> compilation terminated.
> In file included from /tmp/arrayfire/src/backend/common/InteropManager.hpp:13:0,
>                  from /tmp/arrayfire/src/backend/cuda/GraphicsResourceManager.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/platform.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/debug_cuda.hpp:12,
>                  from /tmp/arrayfire/src/backend/cuda/kernel/thrust_sort_by_key_impl.hpp:10,
>                  from /tmp/arrayfire/build/src/backend/cuda/kernel/thrust_sort_by_key/thrust_sort_by_key_impl_float_1.cu:10:
> /tmp/arrayfire/src/backend/common/forge_loader.hpp:14:23: fatal error: glad/glad.h: No such file or directory
> compilation terminated.
> CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o.cmake:207 (message):
>   Error generating
>   /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o
>
>
> CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o.cmake:207 (message):
>   Error generating
>   /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o
>
>
> CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o.cmake:207 (message):
>   Error generating
>   /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o
>
>
> CMake Error at cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o.cmake:207 (message):
>   Error generating
>   /tmp/arrayfire/build/src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/./cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o
>
>
> src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:63: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
> mpl_float_0.cu.o' failed
> make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_0.cu.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:77: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
> mpl_double_0.cu.o' failed
> make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_double_0.cu.o] Error 1
> src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:70: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_$
> mpl_float_1.cu.o' failed
> make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_float_1.cu.o] Error 1
> src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/build.make:196: recipe for target 'src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key$
> impl_uchar_1.cu.o' failed
> make[2]: *** [src/backend/cuda/CMakeFiles/cuda_compile.dir/kernel/thrust_sort_by_key/cuda_compile_generated_thrust_sort_by_key_impl_uchar_1.cu.o] Error 1
> CMakeFiles/Makefile2:1331: recipe for target 'src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/all' failed
> make[1]: *** [src/backend/cuda/CMakeFiles/cuda_thrust_sort_by_key.dir/all] Error 2
> make[1]: *** Waiting for unfinished jobs....
> ...
> [ 12%] Building CXX object src/api/unified/CMakeFiles/af.dir/__/cpp/wrap.cpp.o
> [ 12%] Building CXX object src/api/unified/CMakeFiles/af.dir/__/cpp/ycbcr_rgb.cpp.o
> [ 12%] Linking CXX shared library libaf.so
> [ 12%] Built target af
> Makefile:160: recipe for target 'all' failed
> make: *** [all] Error 2
>
> If I am doing the same but for v3.6.4:
>
> Scanning dependencies of target afcuda
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/kernel/convolve_separable.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/transpose.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/Array.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/blas.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/convolve.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cufft.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cublas.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cusolverDn.cpp.o
> [ 52%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/cusparse.cpp.o
> [ 53%] Building CXX object src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o
> In file included from /tmp/arrayfire/src/backend/cuda/device_manager.cpp:31:0:
> /usr/local/cuda/include/cuda_gl_interop.h:67:19: fatal error: GL/gl.h: No such file or directory
> compilation terminated.
> src/backend/cuda/CMakeFiles/afcuda.dir/build.make:782: recipe for target 'src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o' failed
> make[2]: *** [src/backend/cuda/CMakeFiles/afcuda.dir/device_manager.cpp.o] Error 1
> make[2]: *** Waiting for unfinished jobs....
> CMakeFiles/Makefile2:1403: recipe for target 'src/backend/cuda/CMakeFiles/afcuda.dir/all' failed
> make[1]: *** [src/backend/cuda/CMakeFiles/afcuda.dir/all] Error 2
> Makefile:160: recipe for target 'all' failed
> make: *** [all] Error 2
>
> Thanks for helping!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/arrayfire/arrayfire/issues/2516?email_source=notifications&email_token=AAY6OOQUQZY343JPO46TKULPWM5WZA5CNFSM4HOGL242YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4GU26YYQ>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAY6OOQAURMEG5TOIU5ITULPWM5WZANCNFSM4HOGL24Q>
> .
>

--- a/src/backend/cuda/CMakeLists.txt
+++ b/src/backend/cuda/CMakeLists.txt
@@ -141,6 +141,7 @@ function(cuda_add_library cuda_target)
 endfunction()
 
 arrayfire_get_cuda_cxx_flags(cuda_cxx_flags)
+arrayfire_get_platform_definitions(platform_flags)
 
 if(AF_WITH_NONFREE AND CMAKE_VERSION VERSION_LESS "3.7")
   # This definition is required in addition to the definition below because in
@@ -441,7 +442,7 @@ cuda_add_library(afcuda
 
     nvrtc/cache.cpp
 
-    OPTIONS "${cuda_cxx_flags} -Xcudafe \"--diag_suppress=1427\""
+    OPTIONS "${platform_flags} ${cuda_cxx_flags} -Xcudafe \"--diag_suppress=1427\""
   )
 
 arrayfire_set_default_cxx_flags(afcuda)

--- a/src/backend/cuda/CMakeLists.txt
+++ b/src/backend/cuda/CMakeLists.txt
@@ -35,6 +35,8 @@ mark_as_advanced(
     CUDA_LIBRARIES_PATH
     CUDA_architecture_build_targets)
 
+get_target_property(COMMON_INTERFACE_DIRS afcommon_interface INTERFACE_INCLUDE_DIRECTORIES)
+
 cuda_include_directories(
   ${CMAKE_CURRENT_BINARY_DIR}
   ${CMAKE_CURRENT_SOURCE_DIR}
@@ -46,8 +48,7 @@ cuda_include_directories(
   ${ArrayFire_SOURCE_DIR}/src/api/c
   ${ArrayFire_SOURCE_DIR}/src/backend
 
-  # NOTE: Space after comma is necessary
-  $<JOIN:$<TARGET_PROPERTY:afcommon_interface,INTERFACE_INCLUDE_DIRECTORIES>, >
+  ${COMMON_INTERFACE_DIRS}
   )
 
 set(jit_kernel_headers

